Interactive calibration for Signal2D

magnunor opened the pull request:
This pull requests implements a interactive way of calibrating the signal dimensions in a Signal2D object. It is based on the calibration functionality for Signal1D.

Needs to be used with https://github.com/hyperspy/hyperspy_gui_traitsui/pull/3, or https://github.com/hyperspy/hyperspy_gui_ipywidgets/pull/2

### Progress of the PR
- [x] Initial implementation
- [x] Improve implementation
- [x] Add GUI elements to hyperspy_gui_ipywidgets
- [x] Add docstring to signal2d.calibrate
- [x] Update user guide
- [x] Add tests
- [x] Ready for review.

To test this, grab this branch + the traitsui pull request: https://github.com/hyperspy/hyperspy_gui_traitsui/pull/3.

```python
>>> import hyperspy.api as hs
>>> import numpy as np
>>> s = hs.signals.Signal2D(np.random.random((100, 100)))
>>> s.calibrate(toolkit='traitsui')
```
(This traitsui GUI image is slightly outdated now, since I did some minor changes in the most recent commit).
![calibrate_signal2d](https://user-images.githubusercontent.com/1690979/34157793-52bfeb66-e4c3-11e7-8e49-46b1064ee151.png)

```python
>>> import hyperspy.api as hs
>>> import numpy as np
>>> s = hs.signals.Signal2D(np.random.random((100, 100)))
>>> s.calibrate(toolkit='ipywidgets')
```

![ipywidget_calibrate2d](https://user-images.githubusercontent.com/1690979/34176071-e95753a8-e4fe-11e7-8c4c-0284659e9f5e.png)

francisco-dlp commented on hyperspy/_signals/signal2d.py:
```diff
@@ -599,6 +600,79 @@ def align2D(self, crop=True, fill_value=np.nan, shifts=None, expand=False,
+
+        Parameters
+        ----------
+        x0, y0, x1, y1 : scalars, optional
```
Why not using hyperspy's standard int/float syntax?

francisco-dlp commented on hyperspy/_signals/signal2d.py:
```diff
@@ -599,6 +600,79 @@ def align2D(self, crop=True, fill_value=np.nan, shifts=None, expand=False,
+    def _get_signal2d_scale(self, x0, y0, x1, y1, length, scaled_input=True):
+        sa = self.axes_manager.signal_axes
+        if scaled_input:
+            x0 = x0/sa[0].scale
```
why not using ``DataAxis.value2index``? This may help with consistency.

magnunor replied:
Ended up using `_get_index`, as suggested further down.

francisco-dlp commented on hyperspy/_signals/signal2d.py:
```diff
@@ -599,6 +600,79 @@ def align2D(self, crop=True, fill_value=np.nan, shifts=None, expand=False,
+            y1 = y1/sa[1].scale
+        pos = ((x0, y0), (x1, y1))
+        old_length = np.linalg.norm(np.diff(pos, axis=0), axis=1)[0]
+        scale = length/old_length
```
I think that it'll be good to check if the scale and units of both axes are the same, because otherwise calibrating with a line that is not vertical or horizontal doesn't make sense.

magnunor replied:
Done by adding checks, and outputting a warning if the scale are units are not the same.

francisco-dlp commented on hyperspy/tests/signal/test_2D_tools.py:
```diff
@@ -129,6 +129,96 @@ def test_align_expand(self):
+        assert scale1 == 1.
+
+
+@lazifyTestClass
```
I think that the tests are not running in lazy mode, because in order for that to work, the signals have to be defined in the class constructor e.g.

```python

class TestCalibrate2D:

    def __init__(self):
        self.s = Signal2D()
````

francisco-dlp reviewed (changes requested):
It looks good except for the few comments in the review.

francisco-dlp commented:
I forgot to mention that it'll be good that the line doesn't display the "thickness lines" since they are useless for this purpose.

magnunor commented:
I completely forgot about your comments here @francisco-dlp! I'll have a look at it hopefully soon when I get some time.

francisco-dlp commented:
@magnunor, will you have time to finish this for the v1.4 release (in 1-2 weeks times)?

magnunor commented:
@francisco-dlp, I was busy mountain hiking in Norway the last two weeks, so I haven't had time to look at this...

I _might_ have some time in the evening or this weekend.

magnunor commented:
This should all be fixed now!

Having different `scale` and/or `units` will raise a warning.

Note that https://github.com/hyperspy/hyperspy_gui_ipywidgets/pull/2 and https://github.com/hyperspy/hyperspy_gui_traitsui/pull/3 needs to be merged as well

francisco-dlp commented on hyperspy/_signals/signal2d.py:
```diff
@@ -646,28 +646,53 @@ def calibrate(
+            elif dtype is float:
+                scaled_input = True
+            else:
+                raise TypeError(
```
I think that it should be possible to mix integers and floats here (we do it when indexing and slicing after all). You could simply use ``DataAxis._get_index`` to simplify the code.

magnunor replied:
Done!

francisco-dlp reviewed (changes requested):
In addition, some of the comments of my previous review were not addressed.

magnunor commented:
Everything should be fixed now!

Having the possibility to calibrate the navigation dimensions as well would be nice, but I'll save that for a future pull request.

ericpre commented:
There will be some changes with the matplotlib widgets soon and this tool will most likely benefit from it. I would suggest to revisit after the matplotlib widgets have been improved.

magnunor commented:
Removed this from the 1.7 milestone, as I don't think it will be complete by then.

magnunor commented:
> There will be some changes with the matplotlib widgets soon and this tool will most likely benefit from it. I would suggest to revisit after the matplotlib widgets have been improved.

I haven't really been paying attention to the matplotlib widgets, do you know what the status is on these? @ericpre

ericpre commented:
> > There will be some changes with the matplotlib widgets soon and this tool will most likely benefit from it. I would suggest to revisit after the matplotlib widgets have been improved.
> 
> I haven't really been paying attention to the matplotlib widgets, do you know what the status is on these? @ericpre

This is still work in progress but rather stale...

magnunor commented:
I got some time to look at this, at it seems to work nicely!

It would be nice to include this functionality, even without the reworked matplotlib widgets. This will be an enormous quality of life improvement, so I can dedicate some time for getting it ready for the 1.7 release.

magnunor commented:
I'll cleanup the commit history a bit, as it is currently quite messy

magnunor committed:
Add Signal2D.calibrate function

Includes two new signal_tools:
- LineInSignal2D
- Signal2DCalibration

Also unit tests for the new functionality.

Note that this functionality requires GUI functions implemented in
hyperspy_gui_traitsui or hyperspy_gui_ipywidgets

magnunor committed:
Add user guide documentation for new signal2D.calibrate function

magnunor committed:
Add changelog entry for signal2D.calibrate function (PR 1791)
